src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,808 @@
+import AbstractAlgebra: MatElem, MatSpace, parent_type, Ring, RingElem
+import Hecke: base_ring, det, fq_nmod, FqNmodFiniteField, nrows, tr, trace
+import GAP: FFE
+
+export
+    general_linear_group,
+    matrix_group,
+    MatrixGroup,
+    MatrixGroupElem,
+    omega_group,
+    orthogonal_group,
+    special_linear_group,
+    special_orthogonal_group,
+    special_unitary_group,
+    symplectic_group,
+    unitary_group,
+    GL, GO, GU, SL, SO, Sp, SU
+
+
+
+abstract type AbstractMatrixGroupElem <: GAPGroupElem{GAPGroup} end
+
+"""
+    MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+Type of groups `G` of `n x n` matrices over the ring `R`, where `n = degree(G)` and `R = base_ring(G)`.
+
+At the moment, only rings of type `FqNmodFiniteField` are supported.
+"""
+mutable struct MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+   X::GapObj
+   gens::Vector{<:AbstractMatrixGroupElem}
+   ring::Ring
+   deg::Int
+   descr::Symbol                       # e.g. GL
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+
+   function MatrixGroup(m::Int64, F::S) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      return r
+   end
+
+   function MatrixGroup(m::Int64, F::S, G_gap::GapObj) where S<:Ring
```

Comment:
I would suggest re-using the previous constructor instead of copy-pasting, and maybe not having it as an internal constructor then. E.g.
```julia
function MatrixGroup(m::Int64, F::Ring, G_gap::GapObj)
    M = MatrixGroup(m, F)
    M.X = G_gap
    return M
end
```
Similarly for the third constructor (and for `MatrixGroupElem`). It makes it easier to understand what the differences are between the constructors.

Reply:
So, do I move the lines
`function MatrixGroup(m::Int64, F::Ring, G_gap::GapObj)
    M = MatrixGroup(m, F)
    M.X = G_gap
    return M
end`
outside the struct?

Reply:
Yes. It's generally better to minimize the number of internal constructor (they are meant to enforce internal consistency of the objects; so one or two such constructors are usually enough, and other constructors can rely on them without duplicating the initialization logic).